src/getPage.tsx

```diff
@@ -53,83 +56,85 @@ export default async function getPage({
     router,
     useApp,
     useDocument,
+    nonIsolatedModules,
   };
+
   validateOptions(optionsWithDefaults);
   loadNextConfig({ nextRoot });
-  setNextRuntimeConfig({ runtimeEnv: 'client' });
+  setNextRuntimeConfig({ runtimeEnv: RuntimeEnvironment.CLIENT });
 
   const options: ExtendedOptions = {
     ...optionsWithDefaults,
     pagesDirectory: findPagesDirectory({ nextRoot }),
     pageExtensions: getPageExtensions(),
-    env: 'server',
+    env: RuntimeEnvironment.SERVER,
   };
   // @TODO: Consider printing extended options value behind a debug flag
 
   const headManager = useDocument && initHeadManager();
 
   const makePage = async (
-    optionsOverride?: Partial<ExtendedOptions>
-  ): Promise<Page> => {
-    const mergedOptions = { ...options, ...optionsOverride };
-    let { pageElement, pageData, pageObject } = await makePageElement({
-      options: mergedOptions,
+    options: ExtendedOptions
+  ): Promise<MakePageResult> => {
+    const { pageData, pageObject } = await getPageInfo({ options });
+    const { AppComponent, PageComponent, routeData } = makePageConstructs({
+      pageObject,
+      env: options.env,
     });
 
-    if (useDocument && mergedOptions.env === 'client' && headManager) {
-      pageElement = (
+    const pageElement = (
+      <AppComponent Component={PageComponent} pageProps={pageData.props} />
+    );
+
+    if (
+      !useDocument ||
+      options.env !== RuntimeEnvironment.CLIENT ||
+      !headManager
+    ) {
+      return { pageElement, routeData };
+    }
+
+    return {
+      routeData,
+      pageElement: (
         // @NOTE: implemented from:
         // https://github.com/vercel/next.js/blob/v10.0.3/packages/next/client/index.tsx#L574
         <HeadManagerContext.Provider value={headManager}>
           {pageElement}
         </HeadManagerContext.Provider>
-      );
-    }
-
-    return { pageElement, pageData, pageObject };
+      ),
+    };
   };
 
-  let {
-    pageElement: serverPageElement,
-    pageData,
-    pageObject,
-  } = await makePage();
+  let { pageData, pageObject } = await getPageInfo({ options });
 
-  serverPageElement = (
-    <RouterProvider
-      pageObject={pageObject}
-      options={options}
-      makePage={makePage}
-    >
-      {serverPageElement}
-    </RouterProvider>
-  );
+  const wrapWithRouter = (children: JSX.Element) => {
+    return (
+      <RouterProvider
+        options={options}
+        routeData={pageObject}
+        makePage={(optionsOverrides) =>
+          makePage({ ...options, ...optionsOverrides })
+        }
+      >
+        {children}
+      </RouterProvider>
+    );
+  };
```

Comment:
Could it be moved together with router related stuff?

Reply:
Its hard to move it in current shape as its a closure over a few objects available in scope: `makePage`, `pageObject`, ...

It could be moved if we passed all those to it as arguments.